sky/provision/oci/instance.py

```diff
@@ -0,0 +1,455 @@
+    # VCN (skypilot_vcn). The VCN will only be created at the first
+    # time when it is not existed. We'll not automatically delete the
+    # VCN while teardown clusters. it is intended to let user to decide
+    # to delete the VCN or not from OCI console, for example.
+
+
+@query_utils.debug_enabled(logger)
+def wait_instances(region: str, cluster_name_on_cloud: str,
+                   state: Optional[status_lib.ClusterStatus]) -> None:
+    del region, cluster_name_on_cloud, state
+    # We already wait for the instances to be running in run_instances.
+    # We can not implement the wait logic here because the provisioning
+    # instances are not retrieveable by the QL 'query instance resources ...'.
+
+
+@query_utils.debug_enabled(logger)
+def get_cluster_info(
+    region: str,
+    cluster_name_on_cloud: str,
+    provider_config: Optional[Dict[str, Any]] = None,
+) -> common.ClusterInfo:
+    """Get the metadata of instances in a cluster."""
+    filters = {constants.TAG_RAY_CLUSTER_NAME: cluster_name_on_cloud}
+    running_instances = _get_filtered_nodes(region, filters)
+
+    instances = {}
+    for running_instance in running_instances:
+        inst = _get_inst_obj_with_ip(region, running_instance)
+        instances[inst['id']] = [
+            common.InstanceInfo(
+                instance_id=inst['id'],
+                internal_ip=inst['internal_ip'],
+                external_ip=inst['external_ip'],
+                tags=inst['tags'],
+            )
+        ]
+
+    instances = dict(sorted(instances.items(), key=lambda x: x[0]))
+    logger.debug(f'Cluster info: {instances}')
+
+    head_instance_id = _get_head_instance_id(running_instances)
+    logger.debug(f'Head instance id is {head_instance_id}')
+
+    return common.ClusterInfo(
+        provider_name='oci',
+        head_instance_id=head_instance_id,
+        instances=instances,
+        provider_config=provider_config,
+    )
+
+
+def _get_filtered_nodes(region, tag_filters):
+    return_nodes = []
+
+    try:
+        insts = query_helper.query_instances_by_tags(tag_filters, region)
+    except oci_adaptor.oci.exceptions.ServiceError as e:
+        with ux_utils.print_exception_no_traceback():
+            raise exceptions.ClusterStatusFetchingError(
+                f'Failed to query status for OCI cluster {tag_filters}.'
+                'Details: '
+                f'{common_utils.format_exception(e, use_bracket=True)}')
+
+    for inst in insts:
+        inst_id = inst.identifier
+        return_nodes.append({
+            'inst_id': inst_id,
+            'name': inst.display_name,
+            'ad': inst.availability_domain,
+            'compartment': inst.compartment_id,
+            'status': inst.lifecycle_state,
+            'oci_tags': inst.freeform_tags,
+        })
+
+    return [
+        v for v in return_nodes
+        if v['status'] in ('PROVISIONING', 'STARTING', 'RUNNING', 'STOPPED',
+                           'STOPPING')
```

Comment:
In what case would the status not in this tuple? It will be great to add a comment here about why only keep those status

Reply:
> In what case would the status not in this tuple? It will be great to add a comment here about why only keep those status

We can remove this duplicated checking and safely return the list here because we already excludes the TERMINATING and TERMINATED states in the query logic. Thanks for the comment!